src/hypertable_restrict_info.c

```diff
@@ -0,0 +1,358 @@
+#include <postgres.h>
+#include <nodes/relation.h>
+#include <utils/typcache.h>
+#include <optimizer/clauses.h>
+#include <utils/lsyscache.h>
+#include <parser/parsetree.h>
+
+#include "hypertable_restrict_info.h"
+#include "dimension.h"
+#include "utils.h"
+#include "dimension_slice.h"
+#include "chunk.h"
+#include "dimension_vector.h"
+#include "partitioning.h"
+
+typedef struct DimensionRestrictInfo
+{
+	Dimension  *dimension;
+} DimensionRestrictInfo;
+
+typedef struct DimensionRestrictInfoOpen
+{
+	DimensionRestrictInfo base;
+	int64		lower_bound;	/* internal time representation */
+	StrategyNumber lower_strategy;
+	int64		upper_bound;	/* internal time representation */
+	StrategyNumber upper_strategy;
+} DimensionRestrictInfoOpen;
+
+typedef struct DimensionRestrictInfoClosed
+{
+	DimensionRestrictInfo base;
+	int32		value;			/* hash value */
+	StrategyNumber strategy;	/* either Invalid or equal */
+} DimensionRestrictInfoClosed;
+
+static DimensionRestrictInfoOpen *
+dimension_restrict_info_open_create(Dimension *d)
+{
+	DimensionRestrictInfoOpen *new = palloc(sizeof(DimensionRestrictInfoOpen));
+
+	new->base.dimension = d;
+	new->lower_strategy = InvalidStrategy;
+	new->upper_strategy = InvalidStrategy;
+	return new;
+}
+
+static DimensionRestrictInfoClosed *
+dimension_restrict_info_closed_create(Dimension *d)
+{
+	DimensionRestrictInfoClosed *new = palloc(sizeof(DimensionRestrictInfoClosed));
+
+	new->base.dimension = d;
+	new->strategy = InvalidStrategy;
+	return new;
+}
+
+static DimensionRestrictInfo *
+dimension_restrict_info_create(Dimension *d)
+{
+	switch (d->type)
+	{
+		case DIMENSION_TYPE_OPEN:
+			return (DimensionRestrictInfo *) dimension_restrict_info_open_create(d);
```

Comment:
You can avoid a cast if you return `&dimension_restrict_info_open_create(d)->base`. This is safer in case the `DimensionRestrictInfo` structs change in the future. Same below.